diff --git a/configure.ac b/configure.ac
index da211d0..157b497 100644
--- a/configure.ac
+++ b/configure.ac
@@ -43,6 +43,16 @@ AC_CHECK_HEADERS([pthread.h], [
     AC_MSG_ERROR(The pthread library is required in order to build calcurse!))
 ], AC_MSG_ERROR(The pthread header is required in order to build calcurse!))
 
+AC_CHECK_HEADERS([jalali/jtime.h], [
+    AC_CHECK_LIB(jalali, jstrftime, [
+	LIBS="$LIBS -ljalali"
+	AC_DEFINE(HAVE_LIBJALALI, 1,
+		  [Define to 1 if you have the 'jalali' library (-ljalai).])
+    ],
+    AC_MSG_ERROR(The jalali library is required in order to build jcalcurse!))
+], AC_MSG_ERROR(The jalali header is required in order to build jcalcurse!))
+]
+
 AC_CHECK_HEADERS([math.h], [
     AC_CHECK_LIB(m, exp, [
 	LIBS="$LIBS -lm"
diff --git a/src/calcurse.h b/src/calcurse.h
index 8322416..1c8a828 100644
--- a/src/calcurse.h
+++ b/src/calcurse.h
@@ -1131,6 +1131,7 @@ int string_vcatf(struct string *, const char *, va_list);
 int string_printf(struct string *, const char *, ...);
 int string_catftime(struct string *, const char *, const struct tm *);
 int string_strftime(struct string *, const char *, const struct tm *);
+int string_catf2jtime(struct string *sb, const char *format, struct tm *tm);
 
 /* todo.c */
 extern llist_t todolist;
diff --git a/src/strings.c b/src/strings.c
index ed90bc7..dac9cbd 100644
--- a/src/strings.c
+++ b/src/strings.c
@@ -35,7 +35,10 @@
  */
 
 #include <stdarg.h>
-
+#include <stddef.h>
+#include <time.h>
+#include <jalali/jtime.h>
+#include <jalali/jalali.h>
 #include "calcurse.h"
 
 #define STRING_INITIAL_BUFSIZE 128
@@ -129,6 +132,24 @@ int string_catftime(struct string *sb, const char *format, const struct tm *tm)
 	return n;
 }
 
+int string_catf2jtime(struct string *sb, const char *format, struct tm *tm)
+{
+  struct jtm jtm;
+  time_t t = mktime(tm);
+  jlocaltime_r(&t, &jtm);
+
+	int n = 0;
+
+	while (!n) {
+		string_grow(sb, sb->bufsize * 2);
+		n = jstrftime(sb->buf + sb->len, sb->bufsize - sb->len, format,
+			     &jtm);
+	}
+	sb->len += n;
+
+	return n;
+}
+
 int string_strftime(struct string *sb, const char *format, const struct tm *tm)
 {
 	string_reset(sb);
diff --git a/src/ui-day.c b/src/ui-day.c
index 6a038fa..4ce60ac 100644
--- a/src/ui-day.c
+++ b/src/ui-day.c
@@ -1511,7 +1511,7 @@ static char *fmt_day_heading(time_t date)
 
 	localtime_r(&date, &tm);
 	string_init(&s);
-	string_catftime(&s, conf.day_heading, &tm);
+	string_catf2jtime(&s, conf.day_heading, &tm);
 	return string_buf(&s);
 }
 
