diff --git a/src/ui-calendar.c b/src/ui-calendar.c
index c1719d8..a3fde80 100644
--- a/src/ui-calendar.c
+++ b/src/ui-calendar.c
@@ -505,6 +505,8 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 	t = get_first_weekday(MONDAY);
 	draw_week_number(sw, t);
 
+{
+  struct date slctd_day = ui_calendar_get_slctd_day_jalali();
 	/* Now draw calendar view. */
 	for (j = 0; j < WEEKINDAYS; j++) {
 		/* get next day */
@@ -512,6 +514,9 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 			t = get_first_weekday(wday_start);
 		else
 			date_change(&t, 0, 1);
+    struct jtm jt;
+    time_t time = mktime(&t);
+    jlocaltime_r(&time, &jt);
 
 		struct date date;
 		unsigned attr, item_this_day;
@@ -524,18 +529,18 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 		custom_remove_attr(sw->inner, ATTR_HIGHEST);
 
 		/* Check if the day to be printed has an item or not. */
 		date.dd = t.tm_mday;
 		date.mm = t.tm_mon + 1;
 		date.yyyy = t.tm_year + 1900;
 		item_this_day = day_check_if_item(date);
 
 		/* Print the day numbers with appropriate decoration. */
-		if (t.tm_mday == current_day->dd
+		if (jt.tm_mday == current_day->dd
 		    && current_day->mm == slctd_day.mm
 		    && current_day->yyyy == slctd_day.yyyy
 		    && current_day->dd != slctd_day.dd)
 			attr = ATTR_LOWEST;
-		else if (t.tm_mday == slctd_day.dd)
+		else if (jt.tm_mday == slctd_day.dd)
 			attr = ATTR_MIDDLE;
 		else if (item_this_day == 1)
 			attr = ATTR_LOW;
@@ -548,7 +553,7 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 		if (attr)
 			custom_apply_attr(sw->inner, attr);
 		mvwprintw(sw->inner, OFFY + 1, OFFX + 1 + 4 * j, "%02d",
-			  t.tm_mday);
+			  jt.tm_mday);
 		if (attr)
 			custom_remove_attr(sw->inner, attr);
 		WINS_CALENDAR_UNLOCK;
@@ -568,7 +573,7 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 					int highlight;
 
 					highlight =
-					    (t.tm_mday ==
+					    (jt.tm_mday ==
 					     slctd_day.dd) ? 1 : 0;
 					WINS_CALENDAR_LOCK;
 					if (highlight)
@@ -588,7 +593,7 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 			}
 		}
 	}
-
+}
 	/* Draw marks to indicate midday on the sides of the calendar. */
 	WINS_CALENDAR_LOCK;
 	custom_apply_attr(sw->inner, ATTR_HIGHEST);
