diff --git a/src/ui-calendar.c b/src/ui-calendar.c
index c1719d8..e1a4e90 100644
--- a/src/ui-calendar.c
+++ b/src/ui-calendar.c
@@ -489,8 +489,10 @@ draw_monthly_view(struct scrollwin *sw, struct date *current_day)
 
 /* Draw the weekly view inside calendar panel. */
 static void
-draw_weekly_view(struct scrollwin *sw, struct date *current_day)
+draw_weekly_view(struct scrollwin *sw, struct date *current_day_arg)
 {
+  struct date current_day_value = to_jalali(current_day_arg);
+  struct date *current_day = &current_day_value;
 #define DAYSLICESNO  6
 	const int WCALWIDTH = 28;
 	struct tm t;
@@ -505,6 +507,8 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 	t = get_first_weekday(MONDAY);
 	draw_week_number(sw, t);
 
+{
+  struct date slctd_day = ui_calendar_get_slctd_day_jalali();
 	/* Now draw calendar view. */
 	for (j = 0; j < WEEKINDAYS; j++) {
 		/* get next day */
@@ -512,6 +516,9 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 			t = get_first_weekday(wday_start);
 		else
 			date_change(&t, 0, 1);
+    struct jtm jt;
+    time_t time = mktime(&t);
+    jlocaltime_r(&time, &jt);
 
 		struct date date;
 		unsigned attr, item_this_day;
@@ -530,12 +537,12 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 		item_this_day = day_check_if_item(date);
 
 		/* Print the day numbers with appropriate decoration. */
-		if (t.tm_mday == current_day->dd
+		if (jt.tm_mday == current_day->dd
 		    && current_day->mm == slctd_day.mm
 		    && current_day->yyyy == slctd_day.yyyy
 		    && current_day->dd != slctd_day.dd)
 			attr = ATTR_LOWEST;
-		else if (t.tm_mday == slctd_day.dd)
+		else if (jt.tm_mday == slctd_day.dd)
 			attr = ATTR_MIDDLE;
 		else if (item_this_day == 1)
 			attr = ATTR_LOW;
@@ -548,7 +555,7 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 		if (attr)
 			custom_apply_attr(sw->inner, attr);
 		mvwprintw(sw->inner, OFFY + 1, OFFX + 1 + 4 * j, "%02d",
-			  t.tm_mday);
+			  jt.tm_mday);
 		if (attr)
 			custom_remove_attr(sw->inner, attr);
 		WINS_CALENDAR_UNLOCK;
@@ -568,7 +575,7 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 					int highlight;
 
 					highlight =
-					    (t.tm_mday ==
+					    (jt.tm_mday ==
 					     slctd_day.dd) ? 1 : 0;
 					WINS_CALENDAR_LOCK;
 					if (highlight)
@@ -588,7 +595,7 @@ draw_weekly_view(struct scrollwin *sw, struct date *current_day)
 			}
 		}
 	}
-
+}
 	/* Draw marks to indicate midday on the sides of the calendar. */
 	WINS_CALENDAR_LOCK;
 	custom_apply_attr(sw->inner, ATTR_HIGHEST);
